### ISSUE TITLE
ServiceBusSessionProcessor seems to be able to get session messages out of FIFO order. Is this intentional?

### ISSUE THREAD

**Issue by AnnejanBarelds** [Azure/azure-sdk-for-net#26618]
### Library name and version

Azure.Messaging.ServiceBus 7.5.1

### Query/Question

Hi. I'm currently playing around with Service Bus sessions to get to the bottom of how the FIFO behavior is working. And I'm observing behavior that I wouldn't have expected, so I'm curious if this is intentional.

The setup is quite basic.
- I have a session-enabled queue, and a sender that sends 10 messages to it, all with the same SessionId.
- I have a console app that spins up a ServiceBusSessionProcessor that's not bound to any session in particular.
- AutoCompleteMessages is set to false.
- The ErrorHandler just throws an error.
- The MessageHandler completes all messages, except in case it's message 5 and DeliveryCount is 1. Let's call that the error flow.

What I'm observing is the following:
- In case the error flow is simply to return (i.e. not actually throwing, but also not settling the message), I see that messages 0-4 are being processed as expected, 5 is failing, and then 6-9 are being processed. Then, every minute, message 5 reappears, but without an incremented DeliveryCount. So it keeps triggering the same error flow. Interestingly (and disturbingly) Service Bus Explorer confirms that messages 0-4 and 6-9 are indeed gone, and 5 is still active. So that effectively broke the FIFO ordering. I was not expecting that. It's obviously a programming error to return without settling in this case, but it shouldn't break FIFO ordering in sessions IMHO.
- In case the error flow is to exit the app (Environment.Exit(-1)), I'm observing that

1. The order is preserved: upon restart, I see message 5 again
2. Message 5 does not have an incremented DeliveryCount, however
3. It looks like the lock is abandoned (even though we forcefully exited), because I see message 5 again immediately, as opposed to after a minute or so to account for lock expiration.

- In case the error flow is to throw an error, all works as expected: the message is abandoned and reprocessed, and the DeliveryCount is incremented so it does not trigger the error flow anymore.

So, to sum up, two things happened for which I'm curious if this is intentional:
1. The out-of-FIFO-order delivery of messages in a session if the MessageHandler returns without settling;
2. The fact that DeliveryCount is not incremented on a session message when the MessageHandler either returns without settling, or exits forcefully.

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by JoshLove-msft** [comment]
Will need to check with the service team on the expected behavior here.

**Comment by JoshLove-msft** [comment]
It looks like delivery count works a bit differently when using sessions:
https://docs.microsoft.com/en-us/azure/service-bus-messaging/message-sessions#impact-of-delivery-count

In terms of the FIFO ordering, are you expecting that no messages are returned to you until 5 is settled? This wouldn't work because applications may want to receive multiple messages at a time, so the service does not enforce that a message is settled before returning the next message.

**Comment by AnnejanBarelds** [comment]
Thanks @JoshLove-msft. That accounts for the behavior regarding the delivery count.

As for the FIFO ordering: I'm not sure what you're getting at with your comment that "applications may want to receive multiple messages at a time". In case you're referring to a competing consumer scenario: the whole point of sessions is that only a single consumer can receive messages in a given session at any one time - i.e. the session is locked for that particular consumer.

If you are referring to the possibility to prefetch messages within the context of that single consumer: that may be causing this, but it's still not something I would expect the service to accept. Sessions are the way to guarantee FIFO (https://docs.microsoft.com/en-us/azure/service-bus-messaging/message-sessions#first-in-first-out-fifo-pattern) so if a consumer prefetches messages (which by definition happens in FIFO order) but then settles them out of order, I would expect the service to return an error.

**Comment by JoshLove-msft** [comment]
> If you are referring to the possibility to prefetch messages within the context of that single consumer: that may be causing this, but it's still not something I would expect the service to accept. Sessions are the way to guarantee FIFO (https://docs.microsoft.com/en-us/azure/service-bus-messaging/message-sessions#first-in-first-out-fifo-pattern) so if a consumer prefetches messages (which by definition happens in FIFO order) but then settles them out of order, I would expect the service to return an error.

Assuming you mean prefetch as a concept, and not necessarily using the Prefetch feature of the library, yes this is what I was getting at. The FIFO guarantee is in the delivery of the messages. To my knowledge, there is no enforcement around the order that the messages are settled. That is left up to the consumer to decide.
